subsys/bluetooth/mesh/access.c

```diff
@@ -639,70 +625,74 @@ static int bt_mesh_comp_data_get_page_1(struct net_buf_simple *buf)
 			return -E2BIG;
 		}
 
-		net_buf_simple_add_u8(buf, comp->elem[i].model_count);
-		net_buf_simple_add_u8(buf, comp->elem[i].vnd_model_count);
+		data_buf_add_u8_offset(buf, comp->elem[i].model_count, &offset);
+		data_buf_add_u8_offset(buf, comp->elem[i].vnd_model_count, &offset);
 		for (j = 0; j < comp->elem[i].model_count; j++) {
-			prep_model_item_header(&comp->elem[i].models[j],
-					       &cor_id, &ext_mod_cnt, buf);
+			prep_model_item_header(&comp->elem[i].models[j], &cor_id, &ext_mod_cnt, buf,
+					       &offset);
 			if (ext_mod_cnt != 0) {
-				add_items_to_page(buf,
-						  &comp->elem[i].models[j],
-						  ext_mod_cnt);
+				add_items_to_page(buf, &comp->elem[i].models[j], ext_mod_cnt,
+						  &offset);
 			}
 		}
 
 		for (j = 0; j < comp->elem[i].vnd_model_count; j++) {
-			prep_model_item_header(&comp->elem[i].vnd_models[j],
-					       &cor_id, &ext_mod_cnt, buf);
+			prep_model_item_header(&comp->elem[i].vnd_models[j], &cor_id, &ext_mod_cnt,
+					       buf, &offset);
 			if (ext_mod_cnt != 0) {
-				add_items_to_page(buf,
-						  &comp->elem[i].vnd_models[j],
-						  ext_mod_cnt);
+				add_items_to_page(buf, &comp->elem[i].vnd_models[j], ext_mod_cnt,
+						  &offset);
 			}
 		}
 	}
 	return 0;
 }
 
-static int bt_mesh_comp_data_get_page_2(struct net_buf_simple *buf)
+static int bt_mesh_comp_data_get_page_2(struct net_buf_simple *buf, size_t offset)
 {
 	if (!dev_comp2) {
 		LOG_ERR("Composition data P2 not registered");
 		return -ENODEV;
 	}
 
+	size_t elem_size;
+
 	for (int i = 0; i < dev_comp2->record_cnt; i++) {
-		if (net_buf_simple_tailroom(buf) <
-		    (8 + dev_comp2->record[i].elem_offset_cnt + dev_comp2->record[i].data_len +
-		     BT_MESH_MIC_SHORT)) {
+		elem_size =
+			8 + dev_comp2->record[i].elem_offset_cnt + dev_comp2->record[i].data_len;
+		if (offset >= elem_size) {
+			offset -= elem_size;
+			continue;
+		}
+
+		if (net_buf_simple_tailroom(buf) < ((elem_size - offset) + BT_MESH_MIC_SHORT)) {
 			if (IS_ENABLED(CONFIG_BT_MESH_LARGE_COMP_DATA_SRV)) {
 				/* Mesh Profile 1.1 Section 4.4.1.2.2:
 				 * If the complete list of models does not fit in the Data field,
 				 * the element shall not be reported.
 				 */
-				LOG_DBG("Record 0x%04x didn't fit in the Data field",
-					i);
+				LOG_DBG("Record 0x%04x didn't fit in the Data field", i);
```

Comment:
This function and its siblings are used to store cdp on flash. If LCD is present, but  `CONFIG_BT_MESH_COMP_PST_BUF_SIZE` isn't big enough, it will silently stop here and only part of the cdp will be stored. I think it would make sense to add a check after all CDPs are initialized or in `bt_mesh_comp_store` that would warn if `CONFIG_BT_MESH_COMP_PST_BUF_SIZE` isn't big enough to hold the largest CDP.